optuna/trial/_trial.py

```diff
@@ -63,10 +65,40 @@ def _init_relative_params(self) -> None:
 
         study = pruners._filter_study(self.study, trial)
 
-        self.relative_search_space = self.study.sampler.infer_relative_search_space(study, trial)
-        self.relative_params = self.study.sampler.sample_relative(
-            study, trial, self.relative_search_space
-        )
+        trial_system_attrs = self.system_attrs
+        if "relative_search_space" in trial_system_attrs:
+            self._read_relative_params(study, trial, trial_system_attrs)
+        else:
+            self.relative_search_space = self.study.sampler.infer_relative_search_space(
+                study, trial
+            )
+            self.relative_params = self.study.sampler.sample_relative(
+                study, trial, self.relative_search_space
+            )
+            self._write_relative_search_space_and_params()
+
+    def _write_relative_search_space_and_params(self) -> None:
+        relative_search_space_dict = {
+            k: distribution_to_json(d) for k, d in self.relative_search_space.items()
+        }
+        self.set_system_attr("relative_search_space", relative_search_space_dict)
+        self.set_system_attr("relative_params", self.relative_params)
+
+    def _read_relative_params(
+        self,
+        study: "optuna.study.Study",
+        trial: "optuna.trial.FrozenTrial",
+        trial_system_attrs: Dict[str, Any],
+    ) -> None:
+        relative_search_space = self.study.sampler.infer_relative_search_space(study, trial)
+        relative_search_space_cache_key = "relative_search_space"
+        relative_search_space_from_cache = {
+            k: json_to_distribution(d)
+            for k, d in trial_system_attrs[relative_search_space_cache_key].items()
+        }
+        assert relative_search_space == relative_search_space_from_cache, "search space mismatch"
```

Comment:
Could you explain when this assertion fails?
And also, I'd like to know how users can fix such failure?